android/src/main/java/com/facebook/reactnative/androidsdk/FBAppEventsLoggerModule.java

```diff
@@ -198,6 +200,52 @@ public void logPushNotificationOpen(@Nullable ReadableMap payload) {
          mAppEventLogger.logPushNotificationOpen(Arguments.toBundle(payload));
      }
 
+    /**
+     * Uploads product catalog product item as an app event.
+     *
+     * @param itemID – Unique ID for the item. Can be a variant for a product. Max size is 100.
+     * @param availability – If item is in stock. Accepted values are: in stock - Item ships immediately out of stock - No plan to restock preorder - Available in future available for order - Ships in 1-2 weeks discontinued - Discontinued
+     * @param condition – Product condition: new, refurbished or used.
+     * @param description – Short text describing product. Max size is 5000.
+     * @param imageLink – Link to item image used in ad.
+     * @param link – Link to merchant's site where someone can buy the item.
+     * @param title – Title of item.
+     * @param priceAmount – Amount of purchase, in the currency specified by the 'currency' parameter. This value will be rounded to the thousandths place (e.g., 12.34567 becomes 12.346).
+     * @param currency – Currency used to specify the amount.
+     * @param gtin – Global Trade Item Number including UPC, EAN, JAN and ISBN
+     * @param mpn – Unique manufacture ID for product
+     * @param brand – Name of the brand Note: Either gtin, mpn or brand is required.
+     * @param parameters – Optional fields for deep link specification.
+     */
+    @ReactMethod
+    public void logProductItem(String itemID,
+                               String availability,
+                               String condition,
+                               String description,
+                               String imageLink,
+                               String link,
+                               String title,
+                               BigDecimal priceAmount,
+                               Currency currency,
+                               String gtin,
+                               String mpn,
+                               String brand,
+                               Bundle parameters) {
+        mAppEventLogger.logProductItem(itemID,
+                AppEventsLogger.ProductAvailability.valueOf(availability.toUpperCase()),
+                AppEventsLogger.ProductCondition.valueOf(condition.toUpperCase()),,
```

Comment:
Nice! Thanks for posting this!
I worry a little about these lines for two reasons:

1) if you call String.toUpperCase without a null check, I'm sure there will be crashes with null pointer exceptions later. Should the javascript enforce that these elements are non-nonnull prior to sending them down to native maybe? A follow-on thought is that all of the items may have requirements from the facebook SDK and I didn't see that enforced anywhere other than the types (which are optional in javascript - people may accidentally send in whatever value or undefined...)

2) if you call any of the string mutation operations like toUpperCase without specifying a locale (I suggest `Locale.ROOT` is probably fine?) you can end up with *really* hard to fix errors like this fiendish one https://github.com/invertase/react-native-firebase/issues/3917#issuecomment-843401492

Reply:
Thanks for the quick reply @mikehardy! Heres what i think:

1. Both native SDKs already provide checks for required parameters [here](https://github.com/facebook/facebook-android-sdk/blob/main/facebook-core/src/main/java/com/facebook/appevents/AppEventsLoggerImpl.kt#L198-L231) and [here](https://github.com/facebook/facebook-ios-sdk/blob/989061abaf9e5ec35e7d3d76b38073fa6f60c269/FBSDKCoreKit/FBSDKCoreKit/AppEvents/FBSDKAppEvents.m#L568-L595) so I think we are covered on that end. I only added a null check for those 3 trailing parameters because at least one of them is required, but tbh I'm not sure if throwing an error is even the correct way of handling this case, what do you think?
As for passing in some other type than the one we expect, oh well I guess that's just inherent to using javascript and there's so much we can do. If you really think this type of check would be needed please point me towards the right way of doing it and i'll gladly do it :)

2. Sure thing! I'll _fix_ other usages of the same function too.

Reply:
Hey @santiagofm - re: your reply on point 1, the native SDKs do null check + validate (thanks for the link) so I could agree we are mostly covered there, but if you call toLowerCase on a null string  you will null pointer exception and crash prior to ever hitting the native SDK, so that still needs to be addressed

I agree it's standard javascript that people can send anything, but libraries seem to take two approaches here. Some (like react-native-firebase which I work on) actually validate everything and have a little set of javascript parameter validator utils, and others just trust the developer. fbsdk seems to just trust the developer so I'm fine with continuing to not really validate, but null pointer exceptions must be eliminated as a possibility I think or we will certainly see an issue here later and have to do it anyway

Thanks for throwing the Locale.ROOT in there, I think in general that is important but those errors pop up in very unexpected ways, so nice to eliminate them as a class

Reply:
That's great! I'll source those validation helpers from react-native-firebase !

Reply:
Okay - if you want to do that I think they are all in here and something similar could work https://github.com/invertase/react-native-firebase/blob/master/packages/app/lib/common/validate.js - but as mentioned above I'm not sure it's necessary, I'm most concerned with just making sure we don't actually crash on a NullPointerException in those two specific lines of the java imlementation where toUpperCase are used on possibly null input